test/model/emoji_test.dart

```diff
@@ -0,0 +1,89 @@
+import 'package:checks/checks.dart';
+import 'package:test/scaffolding.dart';
+import 'package:zulip/api/model/model.dart';
+import 'package:zulip/model/emoji.dart';
+
+import '../example_data.dart' as eg;
+
+void main() {
+  group('emojiDisplayFor', () {
+    test('Unicode emoji', () {
+      check(eg.store().emojiDisplayFor(emojiType: ReactionType.unicodeEmoji,
+        emojiCode: '1f642', emojiName: 'smile')
+      ).isA<UnicodeEmojiDisplay>()
+        ..emojiName.equals('smile')
+        ..emojiUnicode.equals('🙂');
+    });
+
+    test('invalid Unicode emoji -> no crash', () {
+      check(eg.store().emojiDisplayFor(emojiType: ReactionType.unicodeEmoji,
+        emojiCode: 'asdf', emojiName: 'invalid')
+      ).isA<TextEmojiDisplay>()
+        .emojiName.equals('invalid');
+    });
+
+    test('realm emoji', () {
+      final store = eg.store(initialSnapshot: eg.initialSnapshot(realmEmoji: {
+        '100': eg.realmEmojiItem(emojiCode: '100', emojiName: 'logo',
+            sourceUrl: '/emoji/100.png'),
+        '123': eg.realmEmojiItem(emojiCode: '123', emojiName: '100',
+            sourceUrl: '/emoji/123.png'),
+        '200': eg.realmEmojiItem(emojiCode: '200', emojiName: 'dancing',
+            sourceUrl: '/emoji/200.png', stillUrl: '/emoji/200-still.png'),
+      }));
+
+      Subject<EmojiDisplay> checkDisplay({
+          required String emojiCode, required String emojiName}) {
+        return check(store.emojiDisplayFor(emojiType: ReactionType.realmEmoji,
+          emojiCode: emojiCode, emojiName: emojiName)
+        )..emojiName.equals(emojiName);
+      }
+
+      checkDisplay(emojiCode: '100', emojiName: 'logo').isA<ImageEmojiDisplay>()
+        ..resolvedUrl.equals(eg.realmUrl.resolve('/emoji/100.png'))
+        ..resolvedStillUrl.isNull();
+
+      // Code matches against code, not against name.
+      checkDisplay(emojiCode: '123', emojiName: '100').isA<ImageEmojiDisplay>()
+        ..resolvedUrl.equals(eg.realmUrl.resolve('/emoji/123.png'))
+        ..resolvedStillUrl.isNull();
```

Comment:
I'm not sure I follow the comment. "Code" (the first word) means "emoji code", right, or does it mean the app code? What bug is this test written against?

Reply:
Yeah, it means "emoji code" as opposed to "emoji name". I'll make the comment less elliptical:
```dart
      // Emoji code matches against emoji code, not against emoji name.
```

The hypothetical bug here isn't a super likely one (because if we had it, then probably most uses would just fail and so we'd notice promptly in manual testing): it's that because both emoji codes and emoji names are just strings, we could take the given emoji name and look it up someplace where the keys are actually supposed to be emoji codes.